Getting Started Documentation

Mluckydwyer opened the pull request:
This PR contains documentation about getting started with Xilinx Co-Simulation with QEMU and SystemCTLM with the included demo. It includes a step-by-step guide for cloning and building the required additional tools (SystemC, Buildroot, and QEMU), running the demo, and sample output a user can expect to see upon completion. This fills a gap in setting up a basic environment that can be further built upon by new users. This also includes `.dtsi` and `.dts` files for describing the hardware described by the demo application for QEMU. They were adapted from the official Xilinx Device Tree Files to work with the test hardware provided in the demo. The README has also been updated to point new users in the direction of the getting started guide and the LMAC demo documentation.

Mluckydwyer committed:
Inital revision of Getting Started Guide documentation on how to build and run the demo applciation

Mluckydwyer committed:
Added modified working .dtsi and .dts files in tandem with Getting Started Documentation to get demo working easily

Mluckydwyer committed:
Updated README to reflect changes in additional documentation

Mluckydwyer committed:
Minor revisions to README formatting

Mluckydwyer commented:
@rc-matthew-l-weber @edgarigl We would love some feedback on these additions to the documentation of the demos. Thank you in advance.

rc-matthew-l-weber commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+# How to set up and run the Co-Simulation Demo
+
+This demonstration shows how to compile and run the Co-Simulation demo of Buildroot in QEMU with a simulated device in SystemC. This configeration is tested working for Ubuntu 18.0.4 and assumes that `cosim` is located in your home directory. This walkthough also assumes the device being emulated is the Zynq-7000 SoC.
```
Using Buildroot to make the OS environment is just one way to set up that part.  For this example, it might be worth noting that Buildroot was selected so that a pure upstream Linux kernel can be built and deployed in QEMU.  This keeps the setup simple and shows how there isn't anything Petalinux specific in creating a complete solution.  (I left you guys to reword this part of it, I proposed some other clarification below for the second half of this intro)

Mluckydwyer replied:
That is a good distinction. All of this should also be compatible with Petalinux or any other env in theory.

rc-matthew-l-weber commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+# How to set up and run the Co-Simulation Demo
+
+This demonstration shows how to compile and run the Co-Simulation demo of Buildroot in QEMU with a simulated device in SystemC. This configeration is tested working for Ubuntu 18.0.4 and assumes that `cosim` is located in your home directory. This walkthough also assumes the device being emulated is the Zynq-7000 SoC.
```
```suggestion
This demonstration shows how to compile and run the Co-Simulation demo of Buildroot in QEMU with a simulated device in SystemC.  This configuration is tested working for Ubuntu 18.0.4 and assumes that a `cosim` directory is created in your home directory. This walkthrough also assumes that the device being emulated by QEMU is the Xilinx Zynq-7000 SoC.  This SoC seemed like a good candidate but the concept can apply to any QEMU machine which plugs in a compatible remoteport bus interface.
```

rc-matthew-l-weber commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+
+# Configure and build
+./configure --target-list="arm-softmmu,aarch64-softmmu,microblazeel-softmmu" --enable-fdt --disable-kvm --disable-xen
+make
```
```suggestion
make -j$((`nproc`+1))
```

Mluckydwyer replied:
This is a good suggestion for ensuring people who copy and paste the commands can take advantage of faster compilation but overall it adds complexity to a demo that is trying to lay everything out in simple terms. Do you think it provides enough benefit to outweigh the complexity it adds?

rc-matthew-l-weber commented on docs/getting-started-guide.md:
````diff
@@ -0,0 +1,168 @@
+mkdir handles
+```
+
+For this walkthough we will be configuring the demo to emulate the Zynq-7000 SoC. The `.dtsi` files available in the Xilinx repositories need significat modification to operate.
````
Not sure I would say that they need a lot of mods like that.  Since you staged the files below, maybe put them in a `support-files` folder and then update this section to pull from that.  Additionally, move the wget's into the headers of the actual dtsi files to show the origin of those files and then comment on what had to change in each respectively.

edgarigl replied:
Thanks, yes, I agree with the comments. It would be nice if we could avoid carrying a copy of the dtsi files in this project.

The idea with the zynq-pl-remote-port.dtsi file in the qemu-devicetrees is to act as an overlay (or an example of an overlay) for a zynq-7000 dts file. If you could provide a diff of the changes you had to do to the dts file, perhaps we can make those changes in the original repository.

Ideally, here we could just do a bunch of wgets to get hold of dts files and a line compile the final dtb (perhaps with a small modification to get zynq-pl-remoteport.dtsi included).

edgarigl replied:
One more thing that may be causing confusion.

QEMU for the Zynq-7000 takes two DTBs as input, -dtb and -hw-dtb.

If only -dtb is passed, that file is used as both the -dtb and -hw-dtb.

-hw-dtb is used by QEMU to instantiate the machine.
-dtb is passed to the ARM Linux kernel.

This means that the -hw-dtb should include the zynq-pl-remoteport.dtsi but the -dtb should not.

On the opposite side, -dtb should include the amba_pl peripherals implemented in the programmable logic (e.g whatever IP you're cosimulating with) so that Linux drivers can bind. The -hw-dtb file should not include those descriptions.

Mluckydwyer replied:
I agree, that would be a better way of including those files. We were just attempting to make it as close to plug and play as possible. I can take a look at the diffs and see how substantial they are.

Mluckydwyer replied:
@edgarigl Are you referring to the [linux-xlnx repo](https://github.com/Xilinx/linux-xlnx) repo or [device-tree-xlnx](https://github.com/Xilinx/device-tree-xlnx)?

edgarigl replied:
Hi,

Ideally, it should work to:
1. wget arch/arm/boot/dts/zynq-zc702.dts from upstream Linux (e.g https://github.com/torvalds/linux.git) or an equivalent dts from the Xilinx repos you mention. Build and use this as -dtb option to QEMU.
2. Remove existing amba_pl nodes from the dts (if any).
3. Add a line at the bottom of the dts so that it includes the zynq-pl-remoteport.dtsi overlay from https://github.com/Xilinx/qemu-devicetrees.git. Use this as the -hw-dtb option to QEMU.

Does that make sense?

Best regards,
Edgar

edgarigl replied:
It's a good point with the plug and play though. An option is to create a script that automates the steps. Perhaps also to include the generated dts files in a sub-directory as Matthew suggested. From my side, I'm open to suggestions on what you guys think is the best way.

rc-matthew-l-weber commented on docs/getting-started-guide.md:
````diff
@@ -0,0 +1,168 @@
+```
+
+## Example Output
+Below is some example output showing the Buildroot instance quererying the SystemC simulated evironment:
````
```suggestion
Below is some example output showing the Buildroot instance querying the SystemC simulated environment:
```

rc-matthew-l-weber commented on docs/getting-started-guide.md:
````diff
@@ -0,0 +1,168 @@
+```bash
+Welcome to Buildroot
+buildroot login: root
+# devmem 0x40000000
````
Maybe describe above what these offsets represent and what you're observing?  i.e., is it a counter?  Additionally, it might be good to add an example of an offset you could write/read back from.

Mluckydwyer replied:
This is a good note. We currently don't really describe these _magic numbers_. If I remember correctly, to write we might need to make further additions but this would be a good addition. I believe we were going for documentation as is, then looking to maybe expand.

rc-matthew-l-weber commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+# How to set up and run the Co-Simulation Demo
```
There are a few demos in this repository, maybe call out the specific one you're show casing.  Then add a ditta/dot diagram(whatever github is most happiest with) that shows how all the parts of the demo come together.  Below that you can then note what tools and purposes each part has.
By adding that visual and outline, the rest of this setup/execute sequence should be easier to follow.  You may even want to breakout the steps and elaborate as you go.

Mluckydwyer replied:
This is a great idea since many people would be able to better visualize what is happening. We are working on some other basic diagrams so we could possibly repurpose those.

edgarigl commented:
> @rc-matthew-l-weber @edgarigl We would love some feedback on these additions to the documentation of the demos. Thank you in advance.

Thanks for working on this! I see that you've already gotten some comments from Matthew. His comments all make sense so I'll wait for a new version of the PR!

Best regards,
Edgar

franciscoIglesias commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+# How to set up and run the Co-Simulation Demo
```
Hi, thanks for this piece of work! Excellent start! I agree with the comments given by @rc-matthew-l-weber and some comments from me are found below. My first comment is to rename this file to include Zynq7000 in the name somehow perhaps as: zynq7000-getting-started-guide.md.

franciscoIglesias commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+
+# Build and install the binaries
+CXXFLAGS=-std=c++11 ./configure ; make
+sudo make install
```
Perhaps not using the the SYSC_VERSION and moving the comment to the top is more clear:

```
# Build and install the binaries
$ cd ~/cosim
$ wget https://www.accellera.org/images/downloads/standards/systemc/systemc-2.3.2.tar.gz
$ tar xf systemc-2.3.2.tar.gz
$ cd systemc-2.3.2/
$ CXXFLAGS=-std=c++11 ./configure --prefix=/usr/local/systemc-2.3.2
$ make
$ sudo make install
```

franciscoIglesias commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+EOF*
+
+# Build the demo
+make
```
In case the zynq_demo is the only one to be built one can do `make zynq_demo`.

franciscoIglesias commented on docs/getting-started-guide.md:
```diff
@@ -0,0 +1,168 @@
+cat << EOF | tee .config.mk
+CXXFLAGS=-std=c++11
+SYSTEMC_INCLUDE=${HOME}/cosim/systemc-2.3.2/src
+SYSTEMC_LIBDIR=${HOME}/cosim/systemc-2.3.2/src/.libs
```
If the `--prefix=/usr/local/systemc-2.3.2/` is used when configuring the systemc-2.3.2  build we should be able to remove above two lines. The Makefile will pick up above as default location through this portion:

```
...
SYSTEMC ?= /usr/local/systemc-2.3.2/
SYSTEMC_INCLUDE ?=$(SYSTEMC)/include/
SYSTEMC_LIBDIR ?= $(SYSTEMC)/lib-linux64
...
```

edgarigl replied:
Installing the systemc libs in ${HOME}/cosim is probably safer as an example since the user may not have access to /usr/local/. But I think we need to include --prefix=${HOME}/cosim/ in the command-line to configure.

Mluckydwyer replied:
@edgarigl I agree, a home install would likely make this walkthrough more accessible.

Mluckydwyer replied:
After a conversation with @rc-matthew-l-weber for the sake of the demo and ease of use, we install in the `--prefix=/usr/local/systemc-2.3.2/` I did add a note about also being able to install it in the home directory if they so choose

franciscoIglesias commented on docs/getting-started-guide.md:
````diff
@@ -0,0 +1,168 @@
+## Running the Demo
+In order to run the co-simulation demo, you will need two shells open. One for QEMU and the other for SystemC. In the first shell navigate to the `systemctlm-cosim-demo` directory (`cd ~/cosim/systemctlm-cosim-demo`). This command will start the SystemC simulation:
+```bash
+LD_LIBRARY_PATH=~/cosim/systemc-2.3.2/src/.libs/ ./zynq_demo 	unix:${HOME}/cosim/buildroot/handles/qemu-rport-_cosim@0 1000000
````
If the `--prefix=/usr/local/systemc-2.3.2/` is used when building systemc-2.3.2. above becomes:

```
LD_LIBRARY_PATH=/usr/local/systemc-2.3.2/lib-linux64/ ./zynq_demo unix:${HOME}/cosim/buildroot/handles/qemu-rport-_cosim@0 1000000
```

franciscoIglesias reviewed (commented):
Hi, thanks for this piece of work! Excellent start! I agree with the comments given by @rc-matthew-l-weber and some comments from me are found below. 

git commit comments: 
squash patch  'Minor revisions to README formatting' into 'Updated README to reflect changes in additional documentation' for obtaining only 1 commit of those 2.

Also in the git commit messages add a summary line and signed-off-by line, similar to the current top of master:

```
libsystemctlm-soc: Bump version

Bump libsystemctlm-soc and update the Makefile as necessary.

Signed-off-by: Edgar E. Iglesias <edgar.iglesias@xilinx.com>
```
Max size of the lines in the message should also be 80 chars (in case it is not possible it will probably also be ok). The signed-of-by line can be obtained by providing the -s switch when creating the commit: `git commit -s`, alternatively it can be added manually.

Great job else!

Thanks!

Best regards,
Francisco

Mluckydwyer committed:
Update docs/getting-started-guide.md

A better description of assumptions made.

Co-authored-by: rc-matthew-l-weber <matthew.weber@rockwellcollins.com>

Mluckydwyer committed:
Update docs/getting-started-guide.md

Typos

Co-authored-by: rc-matthew-l-weber <matthew.weber@rockwellcollins.com>

Mluckydwyer commented:
> Hi, thanks for this piece of work! Excellent start! I agree with the comments given by @rc-matthew-l-weber and some comments from me are found below.
> 
> git commit comments:
> squash patch 'Minor revisions to README formatting' into 'Updated README to reflect changes in additional documentation' for obtaining only 1 commit of those 2.
> 
> Also in the git commit messages add a summary line and signed-off-by line, similar to the current top of master:
> 
> ```
> libsystemctlm-soc: Bump version
> 
> Bump libsystemctlm-soc and update the Makefile as necessary.
> 
> Signed-off-by: Edgar E. Iglesias <edgar.iglesias@xilinx.com>
> ```
> 
> Max size of the lines in the message should also be 80 chars (in case it is not possible it will probably also be ok). The signed-of-by line can be obtained by providing the -s switch when creating the commit: `git commit -s`, alternatively it can be added manually.
> 
> Great job else!
> 
> Thanks!
> 
> Best regards,
> Francisco

Apologies for the non-standard commits. We will fix those when updating with the other suggestions. Thank you for the suggestions.

ctomkins1 commented:
Hi Xilinx team,

Apologies for the lack of communication this past summer, our team had quite the busy summer with internships and other summer activities. The documentation we created last semester should be updated based on your comments and feedback. We are hoping to get this merged in sometime this week if possible. If there are any further updates, please let us know and we can update them to your standards. Thank you!

ctomkins1 committed:
Documentation Updates (#1)

* Updated changes to zynq-7000-getting-started-guide

Updated the documentation for the zynq7000 demo using a system clock.

Signed off by: Cody Tomkins

* Update documentation based on feedback

Minor revisions to documentation. Remove unnecessary bash commands, reword some parragraphs

Signed off by: Cody Tomkins <ctomkins@iastate.edu>

* Update Documentation

Remove unnecessary bash commands, reword paragraphs

Written of by <ctomkins@iastate.edu>

* Delete zynq7000-getting-started-guide.md

Mluckydwyer commented:
The documentation should now have all of those requested changes added except for the diagram mentioned by @rc-matthew-l-weber:

> Then add a ditta/dot diagram(whatever github is most happiest with) that shows how all the parts of the demo come together. Below that you can then note what tools and purposes each part has.

We are working on adding that as well and formalizing the commit.

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
````diff
@@ -0,0 +1,183 @@
+
+```bash
+sudo apt update
+sudo apt install cmake gmake gcc qemu-kvm qemu-system qemu-user-static verilator
````
I tried the commands in the getting started guide using an ubuntu cloud image at: https://cloud-images.ubuntu.com/releases/bionic/release/ubuntu-18.04-server-cloudimg-amd64.img. The packages i required there where this ones:

```
$ sudo apt install gcc verilator build-essential pkg-config libpixman-1-dev libglib2.0-dev expect-dev unzip libssl-dev device-tree-compiler libncurses5-dev libncursesw5-dev
```

For launching a the cloud image:

```
#
# Get the image
#
wget https://cloud-images.ubuntu.com/releases/bionic/release/ubuntu-18.04-server-cloudimg-amd64.img
qemu-img resize ubuntu-18.04-server-cloudimg-amd64.img 20G

#
# Create user-data.img
#
sudo apt-get install cloud-image-utils
cat >user-data <<EOF
#cloud-config
password: pass
chpasswd: { expire: False }
ssh_pwauth: True
EOF

cloud-localds user-data.img user-data


#
# Launch the image
#
qemu-system-x86_64 -M q35,accel=kvm,kernel-irqchip=split -m 4G -smp 4 -device virtio-net-pci,netdev=net0 -netdev type=user,id=net0,hostfwd=tcp::22222-:22 -serial mon:stdio -display none -device intel-iommu,intremap=on,device-iotlb=on -device ioh3420,id=rootport,slot=0 -device ioh3420,id=rootport1,slot=1 -drive file=ubuntu-18.04-server-cloudimg-amd64.img,format=qcow2 -drive file=user-data.img,format=raw

#
# SSH into the image
#
ssh -p 22222 ubuntu@localhost
```

Mluckydwyer replied:
The dependencies that we have listed would then need to be installed on your cloud VM. Are you saying that we should update the packages required to match your list or simply giving instructions on how to replicate your process?
I want to make sure I am understanding you correctly.

franciscoIglesias replied:
Hi Matt,

I apologize for the very unclear comment!

I tested the guide inside the cloud VM as above to try to mimic the situation where a user that hasn't installed much in his system runs the guide and I ended up needing the apt line I provided above (minus 'gcc' since that is a dependency of 'build-essential', installing 'build-essential' installs gcc).

So looking at the command line in the guide at the moment we shouldn't be needing qemu-kvm qemu-system qemu-user-static (since we are building or own qemu). I was not able to 'apt install gmake' in the VM so I'm guessing it might be an alias to the 'cmake' or 'make' package in your environment(?). Its not super easy to say if some packages on my apt line will be brought in by 'cmake' (recursively through it's dependencies, if it does we can keep 'cmake' on the apt line). The easy way out i think is if you try to build the components with the apt command line I provided (minus gcc) and add in more packages in case you need. For testing that out I provided my test setup above in case it could help out (but if you have another setup for testing it is even better. since then two flows will be tested).

Sorry for the long answer, and thank you again for the hard work :)

Best regards,
Francisco Iglesias

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
````diff
@@ -0,0 +1,183 @@
+```bash
+# Build and install the binaries
+cd ~/cosim
+SYSC_VERSION=systemc-2.3.2
````
SYSC_VERSION is not used so we can remove it.

Mluckydwyer replied:
This slipped through the cracks

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+cd ~/cosim
+SYSC_VERSION=systemc-2.3.2
+wget https://www.accellera.org/images/downloads/standards/systemc/systemc-2.3.2.tar.gz
+tar xf systemc.2.3.2.tar.gz
```
s/systemc.2.3.2/system-2.3.2/ (swap first '.' to '-')

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+git clone https://github.com/Xilinx/qemu.git
+cd qemu
+git checkout 7e3e3ae09af8abe70383a5
+git submodule update --init dtc
```
We can remove above two lines (commit 7e3e3ae09af8abe70383a5 could not instantiate one of the nodes in the dtb but master worked fine, and the dtc submodule is cloned automatically when building). An alternative in case we would like to fixate to a specific QEMU commit is to fixate to the current master commit.

Mluckydwyer replied:
We did not see this issue in our testing but I will go ahead and update the fixed commit to the current master. Thank you for the update

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+./configure --target-list="arm-softmmu,aarch64-softmmu,microblazeel-softmmu" --enable-fdt --disable-kvm --disable-xen
+make -j$((`nproc`+1))
+sudo make install
+export SYSTEM=/cosim/systemc-2.3.2/
```
We can remove above line since SYSTEM is not used.

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+git clone https://github.com/Xilinx/systemctlm-cosim-demo.git
+cd systemctlm-cosim-demo
+sed -i 's~../libsystemctlm-soc.git~https://github.com/Xilinx/libsystemctlm-soc.git~g' .gitmodules
+git submodule sync --recursive && git submodule update --init libsystemctlm-soc
```
We can swap above two lines (60, 61) to only:

```
git submodule update --init libsystemctlm-soc
```

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+# Create the Makefile configeration
+cat << EOF | tee .config.mk
+CXXFLAGS=-std=c++11
+SYSTEMC=${HOME/cosim}
```
We need to change above line to:
```
SYSTEMC=${HOME}/cosim/systemc-2.3.2
```

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+CXXFLAGS=-std=c++11
+SYSTEMC=${HOME/cosim}
+HAVE_VERILOG=n
+HAVE_VERILOG_VERILATOR=n
```
This is optional but since verilator was installed above we can do below also:
```
HAVE_VERILOG=y
HAVE_VERILOG_VERILATOR=y
```

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+HAVE_VERILOG_VERILATOR=n
+HAVE_VERILOG_VCS=n
+VM_TRACE=1
+EOF*
```
s/*//

(Remove '*' at the end)

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+# Pull the .dtsi files for the Zynq-7000
+wget https://raw.githubusercontent.com/Xilinx/qemu-devicetrees/master/zynq-pl-remoteport.dtsi
+wget https://raw.githubusercontent.com/Xilinx/linux-xlnx/master/arch/arm/boot/dts/zynq-7000.dtsi
+wget https://raw.githubusercontent.com/Xilinx/device-tree-xlnx/master/device_tree/data/kernel_dtsi/2017.3/BOARD/zc702.dtsi
```
Change above wget to below (for fetching the zynq-zc702.dts and matching the name in BR2_LINUX_KERNEL_CUSTOM_DTS_PATH):
```
# kernel 5.4.75 zynq-zc702.dts
wget https://raw.githubusercontent.com/torvalds/linux/6e97ed6efa701db070da0054b055c085895aba86/arch/arm/boot/dts/zynq-zc702.dts
```
Also after downloading the dts/dtsi files we need to do below for including the remote port setups:
```
echo "#include \"zynq-pl-remoteport.dtsi\"" >> zynq-zc702.dts
```

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+EOF
+
+# Build
+make clean && make olddefconfig
```
Petalinux pre-builts has the default cpufreq governor configured to 'userspace' in the kernel. For configuring this after above line do:

```
$ make linux-menuconfig
    CPU Power Management  --->
        CPU Frequency scaling  --->
            Default CPUFreq governor (ondemand)  --->

                   ***Select userspace***

```
In case it is done after the brmake command one can rebuild the kernel+dtb only in buildroot with 'make linux-rebuild'.

Without this I got a lot of below prints:

```
[    6.379055] cpu cpu0: _generic_set_opp_clk_only: failed to set clock rate: -16
[    6.379083] cpufreq: __target_index: Failed to change cpu frequency: -16
```

Mluckydwyer replied:
If I am understanding this correctly, while a misconfiguration, it does not stop the proper execution of the demo. I am going to look into altering this in an automated fashion when I do my independent verification of the code.

Mluckydwyer replied:
Also, since this is the Xilinx fork of QEMU, does it assume the defaults of PetaLinux and thus this arises since we have chosen Buildroot instead?

franciscoIglesias replied:
Hi Matt,

Its a no on the question above (Xilinx QEMU is not tailored for the petalinux build), I did not look into the issue above further though so not completly sure why the print comes out at the moment. A solution/workaround for the config problem (that worked for me) could be to use the [xiinx_zynq_defconfig](https://github.com/Xilinx/linux-xlnx/blob/master/arch/arm/configs/xilinx_zynq_defconfig) when building the kernel as below:

After the dts/dtsi wget fetches also wget the xiinx_zynq_defconfig inside the ~/cosim/buildroot directory.
```
$ wget https://raw.githubusercontent.com/Xilinx/linux-xlnx/3fbb3a6c57a43342a7daec3bbae2f595c50bc969/arch/arm/configs/xilinx_zynq_defconfig
```

Remove BR2_LINUX_KERNEL_USE_ARCH_DEFAULT_CONFIG=y from the 'cat' command creating the buildroot config and add in below instead into the cat command:
```
$ cat << EOF | tee ".config"
...
BR2_LINUX_KERNEL_USE_CUSTOM_CONFIG=y
BR2_LINUX_KERNEL_CUSTOM_CONFIG_FILE="xilinx_zynq_defconfig"
...
EOF
```
(If you already have another solution mind that is ok aswell!)

Best regards,
Francisco

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
```diff
@@ -0,0 +1,183 @@
+
+# Build
+make clean && make olddefconfig
+./utils/br/make
```
swap to (due to typo):
```
./utils/brmake
```

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
````diff
@@ -0,0 +1,183 @@
+In order to run the co-simulation demo, you will need two shells open. One for QEMU and the other for SystemC. In the first shell navigate to the `systemctlm-cosim-demo` directory (`cd ~/cosim/systemctlm-cosim-demo`). This command will start the SystemC simulation:
+
+```bash
+LD_LIBRARY_PATH=${SYSTEMC}/systemc-2.3.2/src/.libs/ ./zynq_demo 	unix:${HOME}/cosim/buildroot/handles/qemu-rport-_cosim@0 1000000
````
Perhaps creating a variable 'export COSIM=${HOME}/cosim' could simplify but else we can just swap to:
```
LD_LIBRARY_PATH=${HOME}/cosim/systemc-2.3.2/lib-linux64/ ./zynq_demo  unix:${HOME}/cosim/buildroot/handles/qemu-rport-_cosim@0 1000000
```

(SYSTEMC variable has not been setup)

Mluckydwyer replied:
I think the full path is better for demonstration since I agree that setting values can get confusing. I will include the long form.

franciscoIglesias commented on docs/zynq-7000-getting-started-guide.md:
````diff
@@ -0,0 +1,183 @@
+Start the QEMU instance in the second shell by running (in any directory):
+
+```bash
+${SYSTEMC}/qemu/aarch64-softmmu/qemu-system-aarch64 -M arm-generic-fdt-7series -m 1G -kernel ${SYSTEMC}/buildroot/output/images/uImage -dtb ${SYSTEMC}/buildroot/output/images/zynq-zc702.dtb --initrd ${SYSTEMC}/buildroot/output/images/rootfs.cpio.gz -serial /dev/null -serial mon:stdio -display none -net nic -net nic -net user -machine-path ${SYSTEMC}/buildroot/handles -icount 0,sleep=off -rtc cloc ...
````
(The same comment here as above) Below should be ok also:
```
${HOME}/cosim/qemu/aarch64-softmmu/qemu-system-aarch64 -M arm-generic-fdt-7series -m 1G -kernel ${HOME}/cosim/buildroot/output/images/uImage -dtb ${HOME}/cosim/buildroot/output/images/zynq-zc702.dtb --initrd ${HOME}/cosim/buildroot/output/images/rootfs.cpio.gz -serial /dev/null -serial mon:stdio -display none -net nic -net nic -net user -machine-path ${HOME}/cosim/buildroot/handles -icount 0,sleep=off -rtc clock=vm -sync-quantum 1000000

```

franciscoIglesias reviewed (commented):
Hi,

Great work! I provided some comments after trying it out and the last comments are found below! 

In the commit messages: s/Signed off by/Signed-off-by/. If you configure your git user/email the '-s' flag provides it automatically for you:

```
git config --global user.name "firstname lastname"
git config --global user.email  "firstname.lastname@work.com"
git config --global --list
git add file.c
git commit -s     <<< will provide the line 'Signed-off-by: firstname lastname <firstname.lastname@work.com>'
```

After looking I also think that you will only be needing one commit, the one that adds the guide. For removing the other ones you can create the commit on top of current master and force push to the pull request branch (if I understood this [discussion](https://github.com/github/hub/issues/198) correctlly).

Once again, thank you for the excellent work!

Best regards,
Francisco Iglesias

Mluckydwyer committed:
Additional fixes. Verification and kernal config still needed

Signed-off-by: Matthew Dwyer <Mluckydwyer@gmail.com>

Mluckydwyer committed:
Merge branch 'master' of https://github.com/Xilinx/systemctlm-cosim-demo into master

Mluckydwyer committed:
Removed Demo Setup Typo

Mluckydwyer committed:
Fixed dep error

Mluckydwyer committed:
Added wget dep

Mluckydwyer committed:
added g++

Mluckydwyer committed:
Update req packages to match ubuntu 18.04 docker

also fix systemc install prefix

Mluckydwyer committed:
Added buildroot to dep. list for brmake

Mluckydwyer committed:
Create zynq-7000 getting started guide

Create a Zynq-7000 getting started guide in the docs folder to walk a
new user through installing all the needed tools to compile and run a
basic co-simulation demo.

Signed-off-by: Matthew Dwyer <Mluckydwyer@gmail.com>

Mluckydwyer committed:
Add note about SystemC install location

Mluckydwyer committed:
Merge branch 'master' of https://github.com/Mluckydwyer/systemctlm-cosim-demo into master

Mluckydwyer commented:
I have now made all of the additions requested and verified them on a bare Ubuntu 18.04 Docker container (`docker run -it ubuntu:18.04 bash`).

@rc-matthew-l-weber You mentioned being able to assist with the frequency governer part of the Buildroot configuration. Is that still the case?

I am also having some issues squashing all of the commits into one and signing them. It is my understanding though that squash can be done with the merge of the PR, correct?

franciscoIglesias commented:
Hi Matt,

I made some minor final changes (see below), tested the commands and applied the PR (see commit [c7dc0450](https://github.com/Xilinx/systemctlm-cosim-demo/commit/c7dc0450af021053ea0c80b1f80a8a670a66b69f))!

Changes:

* Squashed all commits into one commit
* Made all lines in the doc to be <= 80 chars (for simplifying maintenance)
* Minor modifications on the commands

Thank you for the hard work!

Best regards,
Francisco Iglesias

Mluckydwyer commented:
Hello @franciscoIglesias Thank you  for your assistance, we really appreciate it. Apologies on the styling, that's slipped my mind. That's is great to hear though. Hopefully this can help wrap up this PR that's has been awhile in the making. @rc-matthew-l-weber were you able to make those CPU Frequency Governor changes?

franciscoIglesias commented:
Hi Matt,

No worries and thanks again for the contribution! :)

About the CPU frequency governor printks: I added in the wget for the xilinx_zynq_defconfig along with the buildroot config update to use that defconfig in the guide (instead of using the default for the arch Arm it is better to use the kernel defconfig targeting the zynq). The zynq defconfig has selected the CPU Frequency Governor to userspace already (fixing the printk problem). Basically, I think we can close this PR and in case and you would like to add/change something we can always open a new PR.

Best regards,
Francisco Iglesias

Mluckydwyer commented:
Hello Francisco,

That sounds great. Thank you again for all of your assistance. Those solutions sound warranted and like you mentioned, we can also open another PR if changes need to be made.

Best,
Matthew Dwyer

edgarigl commented:
Merged.
